[native] Add support for Iceberg connector in presto_protocol

imjalpreet opened the pull request:
## Description
<!---Describe your changes in detail-->
Add Java classes from the Iceberg connector in presto_protocol.yml file.
Specialize IcebergColumnHandle to support 'operator<()'.

Depends on #21764 

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->
https://github.com/facebookincubator/velox/issues/5977

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->
With these changes, we will be able to read iceberg tables from Prestissimo

## Test Plan
<!---Please fill in how you tested your change-->
Successfully ran all TPC-H queries with iceberg tables (both partitioned and non-partitioned)

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Add support for Iceberg connector in presto_protocol
```

majetideepak commented:
@imjalpreet is this ready for review?

imjalpreet commented:
@majetideepak Yes the changes are all there in this PR. But need to rebase on master and make some minor updates based on new changes added in master. But I was facing a build error in my local most likely related to https://github.com/prestodb/presto/pull/21797 so I haven't pushed the rebased version of this branch yet.

Once I verify the changes after a successful build, I will push them as well. You can still review the PR since it mostly has all the changes or we can wait for the rebased version which has minor additions.

majetideepak commented:
Why do we need so many `special` changes? Are we not able to translate the Java source files directly?
We should avoid adding to `special` as they are harder to maintain.

yingsu00 commented on presto-native-execution/presto_cpp/main/PrestoServer.cpp:
```diff
@@ -852,6 +853,11 @@ std::vector<std::string> PrestoServer::registerConnectors(
       PRESTO_STARTUP_LOG(INFO) << "Registering catalog " << catalogName
                                << " using connector " << connectorName;
 
+      // TODO: Temporary Change to use HiveDataSource for Iceberg Connector
```
Moving the discussion in #21584 to here:

Add an iceberg.properties in presto-native-execution/etc/catalog with the following
```
# The Presto "iceberg" catalog is handled by the hive connector in Presto native execution.
connector.name=hive

cache.enabled=true
```

imjalpreet replied:
@yingsu00 I have added the catalog file in #21584 since the final changes are part of that PR

yingsu00 commented on presto-native-execution/presto_cpp/presto_protocol/Connectors.cpp:
```diff
@@ -44,6 +44,10 @@ void registerHiveConnectors() {
 }
 
+void registerIcebergConnector() {
+  registerConnector("iceberg", "iceberg");
```
Let's keep this as is, but add a comment saying the connectorName and connectorKey are for Presto, not Presto native.

imjalpreet replied:
@yingsu00 I have made the change as we had discussed earlier and moved the iceberg registration as part of the hive connectors

mbasmanova commented:
> Why do we need so many `special` changes? Are we not able to translate the Java source files directly? We should avoid adding to `special` as they are harder to maintain.

@majetideepak @tdcmeehan Deepak, Tim, do we have any ongoing efforts to switch from JSON to Thrift for communication? JSON-based protocols are pretty challenging to maintain.

tdcmeehan commented:
@mbasmanova  For clarity, there are a few reasons this is not scalable:

1) We must manually update the protocol every time connectors change or there is a new connector.
2) This protocol must be built-in to the Prestissimo binaries.

While 1 is problematic, I believe 2 is the problem that should be solved first (if we must choose between the two), because while 1 is an inconvenience, 2 is more than an inconvenience: it's a blocker from real connector SPI support in C++.

Consider that, to solve 2, we must make it so that connector data structures are not built, but deserialized in a custom way (which is how it works in Java, albeit tied JSON).  The idea is to make it so that connector data structures are serialized in a custom manner, and allow connectors to self-specify how to serialize and deserialize them.  This may include Thrift, but it will be up to the connector how to accomplish that.

When thought of in this way, you're left with two parts to serialize and deserialize: the outer shell of the task update request, and the inner structures used by connectors.  The outer shell may be moved to Thrift.  We would love to get community help to migrate this as well, but I don't think it's the top priority since it changes more slowly and it doesn't block federation use cases in Prestissimo.

mbasmanova commented on presto-native-execution/presto_cpp/presto_protocol/special/BaseHiveColumnHandle.hpp.inc:
```diff
@@ -0,0 +1,30 @@
+ */
+namespace facebook::presto::protocol {
+
+struct BaseHiveColumnHandle : public ColumnHandle {
```
Would you add comments to explain why all these classes cannot be auto-generated?

majetideepak replied:
@mbasmanova I am working with @imjalpreet to try and automate these as well. We will add comments if we are unable to auto-generate.

mbasmanova replied:
@majetideepak Got it. Thank you for sharing. Great to know that you are trying to avoid manual overrides. Thanks.

aditi-pandit commented on presto-native-execution/presto_cpp/presto_protocol/special/ColumnType.cpp.inc:
Can this enum be automated as well ? I feel it should be.

imjalpreet replied:
Oh, yes I missed removing this enum. I had already refactored on the Java side to be able to remove this special class.

imjalpreet replied:
I will remove this in my next push

aditi-pandit reviewed (commented):
We should try to automate the Column(Table)Handle and TableLayout classes.

Are we blocked on them because of the use of inheritance e.g. `struct HiveTableHandle : public BaseHiveTableHandle` ?

imjalpreet commented:
> Are we blocked on them because of the use of inheritance e.g. struct HiveTableHandle : public BaseHiveTableHandle ?

@aditi-pandit Yes, we are trying to see if it would be possible to somehow incorporate multi-level inheritance in these classes (ColumnHandle, TableHandle, and TableLayoutHandle) without the introduction of special classes.

yingsu00 commented:
@majetideepak @mbasmanova @tdcmeehan @aditi-pandit Thanks everyone for chiming in. @imjalpreet and I had some offline discussion today. We think there are two ways to solve this problem:

1. Make presto protocol to support multi-level inheritance (This PR)
In the last a couple of days, @imjalpreet has been trying to modify the Ser/De code to make it support multi-level inheritance and was able to reduce most extra specials. In this approach, presto_protocol.yml was modified to add a "super" member in the subclasses lists in addition to the current "name" and "key" members. We also had to move the `partitionColumns` back from BaseHiveTableLayoutHandle to HiveTableLayoutHandle and IcebergTableLayoutHandle to resolve the problem the `partitionColumns` could not be deserialized because it was List<BaseHiveColumnHandle> and the deserializer doesn't know its actual type.  Jalpreet has updated this PR with this approach.

2. Only serialize leaf classes
    In this approach , BaseHiveTableLayoutHandle and BaseHiveColumnHandle will be removed from presto_protocol, and only HiveTableLayoutHandle, IcebergTableLayoutHandle and HiveColumnHandle, IcebergColumnHandle would be serialized/deserialized. Then we will remove the references of these base classes in presto-native-execution (there is only one real callsite and it can be removed). @imjalpreet can also send a PR on this.

Both approachs need additional work. Before we proceed, we'd like to get your opinions which route we are going to take. IMHO the first approach is more generic and may benefit future changes. But we may need to use a slightly different approach there: instead of adding a "super" member in presto_protocol.yml, I think it's better to list the classes in a separate entry:
```
    ConnectorTableHandle:
      super: JsonEncodedSubclass
      subclasses:
        - { name: BaseHiveTableHandle,      key: base-hive }
        - { name: TpchTableHandle,          key: tpch }
        -
    BaseHiveTableHandle:
      super: JsonEncodedSubclass
      subclasses:
        - { name: HiveTableHandle,          key: hive }
        - { name: IcebergTableHandle,       key: hive-iceberg }
```
This needs additional work in order to build the tree structure. Your opinions on which way to go is very appreciated.

majetideepak commented:
@imjalpreet and @yingsu00 thanks for working on this and finding solutions to avoid manual overrides.
Among the two approaches, I prefer the latter `Only serialize leaf classes` since I see no use in maintaining the inheritance information in the protocol. 
In the future, we can add support for' super' when there is a use-case for the inheritance details.

> Both approachs need additional work.

Can you share what additional work is required for the second approach?

imjalpreet commented:
@majetideepak I have updated the PR to remove the usage of Base classes and excluded inheritance in the protocol. 

Based on these changes I have also updated the implementation in the follow-up PR: https://github.com/prestodb/presto/pull/21584

> Can you share what additional work is required for the second approach?
I think Ying was referring to the implementation changes for the current PRs. There are no changes required in protocol generation scripts like the inheritance approach.

Please have a look at the updated changes and let me know if these look good or in case there are any questions.

yingsu00 commented:
@majetideepak Jalpreet has created a PR for each way:

1. Multiple inheritance: [[iceberg] [native] Introduce Iceberg Connector in Prestissimo (with multi-level inheritance)](https://github.com/prestodb/presto/pull/21905#top)
2. Only leaf classes: This PR.

majetideepak reviewed (approved):
@imjalpreet Thank you for extending the protocol without a lot of specialization.
Just to confirm, the `presto_protocol.h` and `presto_protocol.cpp` files have not been manually edited correct?

majetideepak reviewed (changes requested):
@imjalpreet please fix the commit title and description. Thanks!

mbasmanova reviewed (commented):
@imjalpreet This version is looking nice, clean and simple. Curious, what was the trick that allowed to remove the specializations.

imjalpreet commented:
> Thank you for extending the protocol without a lot of specialization. Can you confirm the `presto_protocol.h` and `presto_protocol.cpp` files have not been manually edited?

@majetideepak yes, the files have not been manually edited. 

> please fix the commit title and description. Thanks!

Can you please let me know what would you like to fix in the commit message?

imjalpreet commented:
> This version is looking nice, clean and simple. Curious, what was the trick that allowed to remove the specializations.

@mbasmanova There are some enums from the iceberg library that are required in the protocol. We have created wrappers for them on the java side so that they can be automated using the yaml file. Apart from that, earlier we were trying to replicate the multi-level inheritance present on the java side in prestissimo as well. But since currently we don't have a big usage of the inheritance details in Prestissimo, we have updated our implementation to remove their usage in Prestissimo. If in the future, we see that inheritance will be useful, we can have a look at https://github.com/prestodb/presto/pull/21905 for one of the ways we can update the protocol generation process to simplify the inclusion of multi-level inheritance.

mbasmanova commented:
@imjalpreet Got it. Thank you for explaining.

majetideepak commented:
@imjalpreet can you add the description to the commit body as well? You need to add an empty line after the title and add the body.
See example below.

```
commit 1bd7843cd321023d04c6cba734b136b6c552ea96
Author: Deepak Majeti <deepak.majeti@ibm.com>
Date:   Mon Feb 5 01:49:44 2024 -0500

    [native] Remove RE2 dependency from centos setup script
    
    RE2 is installed in the Velox centos setup script
```

majetideepak commented:
Just these 2 lines in the commit body.
```
Add Java classes from the Iceberg connector in presto_protocol.yml file.
Specialize IcebergColumnHandle to support 'operator<()'.
```

majetideepak reviewed (approved):
Thanks, @imjalpreet

yingsu00 commented on presto-native-execution/presto_cpp/presto_protocol/presto_protocol.yml:
```diff
@@ -177,6 +181,7 @@ JavaClasses:
   - presto-hive-metastore/src/main/java/com/facebook/presto/hive/BucketFunctionType.java
   - presto-hive-common/src/main/java/com/facebook/presto/hive/CacheQuotaRequirement.java
   - presto-hive-common/src/main/java/com/facebook/presto/hive/CacheQuotaScope.java
+  - presto-hive-common/src/main/java/com/facebook/presto/hive/BaseHiveColumnHandle.java
```
Is this still needed?

imjalpreet replied:
Yes, this is still needed. ColumnType enum is part of this class and it is required to be included in the protocol.

yingsu00 reviewed (commented):
@imjalpreet 
I see that the Json constructor for BaseHiveTableLayoutHandle is still there. Is it still needed? I don't see it's being used. Can you please add a separate commit or PR to remove it If it's no longer needed? 

BaseHiveColumnHandle  fields are still json annotated. Are the annotations required?

imjalpreet commented on presto-native-execution/presto_cpp/presto_protocol/special/HiveColumnHandle.hpp.inc:
```diff
@@ -12,6 +12,9 @@
  * limitations under the License.
  */
 
+// HiveColumnHandle is special since we require an implementation of
```
Sure, I will make this change.

imjalpreet replied:
I updated the comment

imjalpreet commented:
> I see that the Json constructor for BaseHiveTableLayoutHandle is still there. Is it still needed? I don't see it's being used. Can you please add a separate commit or PR to remove it If it's no longer needed? 
> 
> 
> 
> BaseHiveColumnHandle  fields are still json annotated. Are the annotations required? 

@yingsu00 Since, this PR only includes protocol changes I haven't refactored any presto classes here.

I will have a look at the Base* classes in Presto, verify the need for JSON constructors and refactor them as needed in a separate PR.
